src/createAnimatedComponent/NativeEventsManager.ts

```diff
@@ -0,0 +1,118 @@
+'use strict';
+import type {
+  INativeEventsManager,
+  IAnimatedComponentInternal,
+  AnimatedComponentProps,
+  InitialComponentProps,
+} from './commonTypes';
+import { has } from './utils';
+import { WorkletEventHandler } from '../reanimated2/WorkletEventHandler';
+import { isWeb } from '../reanimated2/PlatformChecker';
+
+const IS_WEB = isWeb();
+
+type ManagedAnimatedComponent = React.Component<
+  AnimatedComponentProps<InitialComponentProps>
+> &
+  IAnimatedComponentInternal;
+
+type WorkletEventHandlerProp = {
+  workletEventHandler: InstanceType<typeof WorkletEventHandler>;
+};
+
+function isWorkletEventHandler(prop: unknown): prop is WorkletEventHandlerProp {
+  return (
+    has('workletEventHandler', prop) &&
+    prop.workletEventHandler instanceof WorkletEventHandler
+  );
+}
+
+function executeForEachEventHandler(
+  props: AnimatedComponentProps<InitialComponentProps>,
+  callback: (
+    key: string,
+    handler: InstanceType<typeof WorkletEventHandler>
+  ) => void
+) {
+  for (const key in props) {
+    const prop = props[key];
+    if (isWorkletEventHandler(prop)) {
+      callback(key, prop.workletEventHandler);
+    }
+  }
+}
+
+export class NativeEventsManager implements INativeEventsManager {
+  _managedComponent: ManagedAnimatedComponent;
+
+  constructor(component: ManagedAnimatedComponent) {
+    this._managedComponent = component;
+  }
+
+  public attachNativeEvents(): void {
+    if (IS_WEB) {
+      return;
+    }
+    executeForEachEventHandler(this._managedComponent.props, (key, handler) => {
+      handler.registerForEvents(this._managedComponent._eventViewTag, key);
+    });
+  }
+
+  public detachNativeEvents(): void {
+    if (IS_WEB) {
+      return;
+    }
+    executeForEachEventHandler(
+      this._managedComponent.props,
+      (_key, handler) => {
+        handler.unregisterFromEvents(this._managedComponent._eventViewTag);
+      }
+    );
+  }
+
+  public updateNativeEvents(
+    prevProps: AnimatedComponentProps<InitialComponentProps>,
+    computedEventTag: number
+  ): void {
+    if (IS_WEB) {
+      return;
+    }
+    // If the event view tag changes, we need to completely re-mount all events
+    if (this._managedComponent._eventViewTag !== computedEventTag) {
+      // Remove all bindings from previous props that ran on the old viewTag
+      executeForEachEventHandler(prevProps, (_key, handler) => {
+        handler.unregisterFromEvents(this._managedComponent._eventViewTag);
+      });
+      // We don't need to unregister from current (new) props, because their events weren't registered yet
+      // Replace the view tag
+      this._managedComponent._eventViewTag = computedEventTag;
```

Comment:
I don't like the fact that this is where the `eventViewTag` is set but it's calculated in `createAnimatedComponent`. It's looks like it could be a confusing side effect in the future. Let's move all the responsibility for handling event tags to this manager. We should have a method `updateEventTag` that would be called inside `updateNativeEvents`.